File: modules/share-intent/ios/ShareIntentModule.swift
import ExpoModulesCore

/**
 A share handed to Gryt by another app.

 iOS does not deliver a share to the app. It delivers it to a **Share
 Extension** — a separate process, launched inside the sending app, which has no
 way to talk to Gryt directly. The road between them is the same App Group
 container the ReplayKit extension uses: the extension copies what was shared
 into it, writes a small manifest naming the copies, and opens Gryt. This reads
 that manifest.

 So there is nothing to receive until the extension exists. Until then `consume`
 returns null on every call, which is also what it returns on the vast majority
 of calls afterwards — an ordinary launch has no share waiting.

 **The copies are the extension's, and the cleanup is ours.** A share extension
 is killed the moment it finishes; it cannot wait for an upload. So it copies
 first and leaves, and everything it left behind is swept the next time a share
 is consumed. Sweeping on read rather than on send means a share somebody
 abandoned still gets cleared, which a send-side cleanup would miss.
 */
public final class ShareIntentModule: Module {
  /** Everything the extension writes lives under one directory in the group. */
  private static let inbox = "share"
  private static let manifest = "pending.json"

  public func definition() -> ModuleDefinition {
    Name("ShareIntent")

    /* Declared so the JavaScript side can be written once for both platforms.
       Nothing sends it here: iOS brings the app to the front to deliver a
       share, and a foreground is the signal. Android has a case where the app
       is already in front, and does send it. */
    Events("onShare")

    Function("consume") { () -> [String: Any]? in
      Self.consume()
    }
  }

  private static func consume() -> [String: Any]? {
    guard let inboxURL = Self.inboxURL else { return nil }
    let manifestURL = inboxURL.appendingPathComponent(Self.manifest)

    guard
      let data = try? Data(contentsOf: manifestURL),
      let raw = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    else {
      /* No manifest is the ordinary case. Sweep anyway: an extension that was
         killed between copying and writing its manifest leaves files nothing
         will ever claim. */
      sweep(inboxURL, keeping: nil)
      return nil
    }

    /* Taken before anything else can fail, so a share that cannot be read is
       not offered again on every foreground for the life of the app. */
    try? FileManager.default.removeItem(at: manifestURL)

    let batch = raw["batch"] as? String
    sweep(inboxURL, keeping: batch)

    let files = (raw["files"] as? [[String: Any]] ?? []).compactMap { entry -> [String: Any]? in
      guard let relative = entry["path"] as? String else { return nil }
      let url = inboxURL.appendingPathComponent(relative)
      guard FileManager.default.fileExists(atPath: url.path) else { return nil }
      return [
        /* A `file://` URL rather than a path, because that is what `fetch` takes
           in the upload — the same thing the image picker hands back. */
        "uri": url.absoluteString,
        "name": entry["name"] as Any,
        "mime": entry["mime"] as Any,
      ]
    }

    let text = raw["text"] as? String
    if files.isEmpty && (text?.isEmpty ?? true) { return nil }

    return ["text": text as Any, "files": files]
  }

  /**
   Delete every batch except the one just claimed.

   Batches are directories named by the extension. Anything that is not the
   current one belongs to a share that was already dealt with, or to one
   somebody backed out of — either way nothing is going to ask for it again, and
   a shared video left in the container is a video taking up the phone's storage
   under a name its owner will never find.
   */
  private static func sweep(_ inbox: URL, keeping batch: String?) {
    let contents = try? FileManager.default.contentsOfDirectory(
      at: inbox,
      includingPropertiesForKeys: nil,
    )
    for url in contents ?? [] {
      if url.lastPathComponent == Self.manifest { continue }
      if let batch, url.lastPathComponent == batch { continue }
      try? FileManager.default.removeItem(at: url)
    }
  }

  /**
   The shared container, or nil where there is not one.

   `RTCAppGroupIdentifier` is the authority: it is written into Info.plist by
   `plugins/withScreenShare.js`, which is also what registers the group in the
   first place, so reading it here means one source of truth rather than the
   same string typed in two config plugins. The name is the WebRTC library's
   because that is what first needed a group; the container is the app's.
   */
  private static var inboxURL: URL? {
    guard
      let identifier = Bundle.main.object(forInfoDictionaryKey: "RTCAppGroupIdentifier") as? String,
      let container = FileManager.default.containerURL(
        forSecurityApplicationGroupIdentifier: identifier,
      )
    else { return nil }

    let inbox = container.appendingPathComponent(Self.inbox)
    guard FileManager.default.fileExists(atPath: inbox.path) else { return nil }
    return inbox
  }
}

File: modules/share-intent/android/src/main/java/chat/gryt/shareintent/ShareIntentModule.kt
package chat.gryt.shareintent

import android.content.ClipDescription
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.provider.OpenableColumns
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * A share handed to Gryt by another app.
 *
 * Android delivers this as an `Intent` — `ACTION_SEND` with one `EXTRA_STREAM`,
 * or `ACTION_SEND_MULTIPLE` with several, or `EXTRA_TEXT` for a link. None of
 * that reaches JavaScript on its own: `Linking` only sees a URL, and a share is
 * not a URL. So this reads the Intent and flattens it into the same shape the
 * iOS side produces.
 *
 * **A share is consumed once.** The launch Intent stays on the Activity for as
 * long as the Activity lives, so reading it without clearing it means the same
 * picture is offered again on every foreground until the app is killed —
 * indistinguishable, from the inside, from somebody sharing the same picture
 * repeatedly. `consumed` is what stops that.
 */
class ShareIntentModule : Module() {
  /**
   * The Intent to read, when it arrived while the app was already running.
   *
   * `onNewIntent` does not replace what `getIntent()` returns unless somebody
   * calls `setIntent`, and relying on the host Activity to have done that is
   * relying on a detail of a class this module does not own. Holding it here is
   * one field and no assumptions.
   */
  private var pending: Intent? = null
  private var consumed = false

  override fun definition() = ModuleDefinition {
    Name("ShareIntent")

    Events("onShare")

    /**
     * The share waiting to be dealt with, or null.
     *
     * Null is the ordinary answer. Every cold start reads its launch Intent
     * through here whether or not it was a share, and almost none of them were.
     */
    Function("consume") { ->
      val intent = pending ?: appContext.currentActivity?.intent
      if (intent == null || consumed) return@Function null

      val share = read(intent)
      /* Marked whether or not there was anything in it — an ordinary launch is
         consumed too, so a later foreground does not re-read an Intent that has
         already been looked at. */
      consumed = true
      pending = null
      share
    }

    OnNewIntent { intent ->
      /* A second share while the app is already up. The Activity is coming to
         the front either way, so the event is not what brings the sheet up —
         it is what tells JavaScript there is something new behind the
         foreground it was about to see anyway. */
      if (!isShare(intent)) return@OnNewIntent
      pending = intent
      consumed = false
      sendEvent("onShare", mapOf("waiting" to true))
    }
  }

  private fun isShare(intent: Intent): Boolean =
    intent.action == Intent.ACTION_SEND || intent.action == Intent.ACTION_SEND_MULTIPLE

  private fun read(intent: Intent): Map<String, Any?>? {
    if (!isShare(intent)) return null

    val uris = mutableListOf<Uri>()
    if (intent.action == Intent.ACTION_SEND) {
      streamOf(intent)?.let { uris.add(it) }
    } else {
      streamsOf(intent)?.let { uris.addAll(it) }
    }

    /* `EXTRA_TEXT` is the link or the note. It also arrives *alongside* files
       from apps that add "Shared from …", which is why both halves are read
       rather than one being treated as the other's absence. */
    val text = intent.getCharSequenceExtra(Intent.EXTRA_TEXT)?.toString()
    val subject = intent.getStringExtra(Intent.EXTRA_SUBJECT)

    val files = uris.map { describe(it, intent.type) }
    if (files.isEmpty() && text.isNullOrBlank() && subject.isNullOrBlank()) return null

    return mapOf(
      /* The subject only when there is no body. Several mail and browser apps
         send the page title as the subject and the URL as the text, and showing
         both would put the title in the message twice. */
      "text" to (if (!text.isNullOrBlank()) text else subject),
      "files" to files,
    )
  }

  @Suppress("DEPRECATION")
  private fun streamOf(intent: Intent): Uri? =
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
      intent.getParcelableExtra(Intent.EXTRA_STREAM, Uri::class.java)
    } else {
      intent.getParcelableExtra(Intent.EXTRA_STREAM)
    }

  @Suppress("DEPRECATION")
  private fun streamsOf(intent: Intent): List<Uri>? =
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
      intent.getParcelableArrayListExtra(Intent.EXTRA_STREAM, Uri::class.java)
    } else {
      intent.getParcelableArrayListExtra(Intent.EXTRA_STREAM)
    }

  /**
   * A name and a type for one `content://` uri.
   *
   * Both are worth the query. The uri itself usually carries neither — an
   * Android media uri is an opaque row id — and without them the upload is
   * stored as `application/octet-stream` under a made-up name. The JavaScript
   * side guesses when this comes back empty, but a real answer beats a guess.
   *
   * Failures here are ordinary rather than exceptional: the sending app grants
   * read permission on the uri only, and asking a provider that has gone away
   * throws. A file with no name is still a file worth sending.
   */
  private fun describe(uri: Uri, fallbackType: String?): Map<String, Any?> {
    val resolver = appContext.reactContext?.contentResolver
    var name: String? = null

    try {
      resolver?.query(uri, arrayOf(OpenableColumns.DISPLAY_NAME), null, null, null)?.use { cursor ->
        val column = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
        if (column >= 0 && cursor.moveToFirst()) name = cursor.getString(column)
      }
    } catch (_: Exception) {
      // Left null. The name is a nicety; the uri is the file.
    }

    val mime = try {
      resolver?.getType(uri)
    } catch (_: Exception) {
      null
    }

    return mapOf(
      "uri" to uri.toString(),
      "name" to name,
      // The Intent's own type is the fallback, but only when it names one thing.
      // A multi-share sets it to a wildcard like "image/" plus a star, which is
      // not a mime type anything can store. (Written out because Kotlin nests
      // block comments, so the literal wildcard opens one that never closes.)
      "mime" to (mime ?: fallbackType?.takeUnless { it.contains('*') || it == ClipDescription.MIMETYPE_TEXT_PLAIN }),
    )
  }
}
